sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -906,19 +955,110 @@ public void setValueDeserializer(String valueDeserializer) {
       Coder<K> keyCoder = getKeyCoder(coderRegistry);
       Coder<V> valueCoder = getValueCoder(coderRegistry);
 
-      // Handles unbounded source to bounded conversion if maxNumRecords or maxReadTime is set.
-      Unbounded<KafkaRecord<K, V>> unbounded =
-          org.apache.beam.sdk.io.Read.from(
-              toBuilder().setKeyCoder(keyCoder).setValueCoder(valueCoder).build().makeSource());
+      if (!isUseSDFTransform()
+          || !ExperimentalOptions.hasExperiment(input.getPipeline().getOptions(), "beam_fn_api")
```

Comment:
It looks that we depend on some specific pipeline business logic here. I'd prefer to avoid this if possible.

Reply:
SDF is only supported over `beam_fn_api`. We shouldn't expand the `KafkaIO.Read` with SDF when the `beam_fn_api` is not enbaled, or `beam_fn_api_use_deprecated_read ` is enabled.